<center><h1> Capstone Project: Predict Diabetes with Watson Machine Learning</h1></center>

## Use Case and Business Requirements

Based on the historical diabetics Diagnostic data to create a prediction model which provides prediction for people who are likely to develop diabetes with 70% or greater accuracy. Deploy the model to Watson Machine Learning for the Model Evaluation 


## Overall Pipeline Steps

The learning goals of this notebook are:

* Load a CSV file into the Object Storage Service linked to Watson Studio
* Pandas Datafame to clean and fix the dataset
* Apache® Spark machine learning model Pipe line with Random Forest Regression
* Train and evaluate the model
* Persist & Delopy a model in a Watson Machine Learning repository
* Call the Watson Machine Learning Web Services 

## Architecture Decisions: Choose Dataset and IBM Services & Tools

* Dataset: https://www.kaggle.com/uciml/pima-indians-diabetes-database (Suggest to download the modified dataset fitting this project from Github) 
* Watson Studio for ETL & Model Development (Jupyter noteBook)
* Watson Machine Learning for Production Deployment (API available for application) 

## 1. Load and explore data

In [460]:
# The code was removed by Watson Studio for sharing.

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  Skin  \
0            6      148             72             35        0  33.6    35   
1            1       85             66             29        0  26.6    29   
2            8      183             64              0        0  23.3     0   
3            1       89             66             23       94  28.1    23   
4            0      137             40             35      168  43.1    35   
5            5      116             74              0        0  25.6     0   
6            3       78             50             32       88  31.0    32   
7           10      115              0              0        0  35.3     0   
8            2      197             70             45      543  30.5    45   
9            8      125             96              0        0   0.0     0   

   DiabetesPedigreeFunction  Age DiabetesDiagnosis  
0                     0.627   50               yes  
1                     0.351   31                no  
2                     0.672   32               yes  
3                     0.167   21                no  
4                     2.288   33               yes  
5                     0.201   30                no  
6                     0.248   26               yes  
7                     0.134   29                no  
8                     0.158   53               yes  
9                     0.232   54               yes

Explore the loaded data by using the following Apache® Spark DataFrame methods:
* print schema
* print top ten records
* count all records

In [461]:
df_pandas.shape

(769, 10)

As you can see, the data contains ten  fields. The  HEARTFAILURE field is the one we would like to predict (label).

In [462]:
df_pandas.isnull().values.any()

False

In [463]:
df_pandas.count()

Pregnancies                 769
Glucose                     769
BloodPressure               769
SkinThickness               769
Insulin                     769
BMI                         769
Skin                        769
DiabetesPedigreeFunction    769
Age                         769
DiabetesDiagnosis           769
dtype: int64

As you can see, the data set contains 769 records.

## 2. Cleanse and fix the Data

* Replace the cells with value 0 to the mean of the column (alternative way is to delete the related rows)
* Remove the columns which is highly correlated with other columns

In [464]:
import numpy
# mark zero values as missing or NaN
df_pandas = df_pandas.replace(0, numpy.NaN)
# fill missing values with mean column values
df_pandas.fillna(df_pandas.mean(), inplace=True)
# count the number of NaN values in each column
print(df_pandas.isnull().sum())

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
Skin                        0
DiabetesPedigreeFunction    0
Age                         0
DiabetesDiagnosis           0
dtype: int64


If indicated by the installer, restart the kernel and rerun the notebook until here and continue with the workshop.

In [465]:
df_pandas.head(10)

Pregnancies  Glucose  BloodPressure  SkinThickness     Insulin        BMI  \
0     6.000000    148.0      72.000000      35.000000  155.379747  33.600000   
1     1.000000     85.0      66.000000      29.000000  155.379747  26.600000   
2     8.000000    183.0      64.000000      29.164207  155.379747  23.300000   
3     1.000000     89.0      66.000000      23.000000   94.000000  28.100000   
4     4.490881    137.0      40.000000      35.000000  168.000000  43.100000   
5     5.000000    116.0      74.000000      29.164207  155.379747  25.600000   
6     3.000000     78.0      50.000000      32.000000   88.000000  31.000000   
7    10.000000    115.0      72.384196      29.164207  155.379747  35.300000   
8     2.000000    197.0      70.000000      45.000000  543.000000  30.500000   
9     8.000000    125.0      96.000000      29.164207  155.379747  32.443404   

        Skin  DiabetesPedigreeFunction  Age DiabetesDiagnosis  
0  35.000000                     0.627   50               yes  
1  29.000000                     0.351   31                no  
2  29.164207                     0.672   32               yes  
3  23.000000                     0.167   21                no  
4  35.000000                     2.288   33               yes  
5  29.164207                     0.201   30                no  
6  32.000000                     0.248   26               yes  
7  29.164207                     0.134   29                no  
8  45.000000                     0.158   53               yes  
9  29.164207                     0.232   54               yes

In [466]:
df_pandas.corr()

Pregnancies   Glucose  BloodPressure  SkinThickness  \
Pregnancies                  1.000000  0.154972       0.260083       0.131059   
Glucose                      0.154972  1.000000       0.219305       0.192242   
BloodPressure                0.260083  0.219305       1.000000       0.191525   
SkinThickness                0.131059  0.192242       0.191525       1.000000   
Insulin                      0.068677  0.420629       0.073559       0.157349   
BMI                          0.111976  0.232015       0.283093       0.540005   
Skin                         0.131059  0.192242       0.191525       1.000000   
DiabetesPedigreeFunction    -0.004637  0.137858      -0.001156       0.099964   
Age                          0.512118  0.267034       0.325326       0.127312   

                           Insulin       BMI      Skin  \
Pregnancies               0.068677  0.111976  0.131059   
Glucose                   0.420629  0.232015  0.192242   
BloodPressure             0.073559  0.283093  0.191525   
SkinThickness             0.157349  0.540005  1.000000   
Insulin                   1.000000  0.167927  0.157349   
BMI                       0.167927  1.000000  0.540005   
Skin                      0.157349  0.540005  1.000000   
DiabetesPedigreeFunction  0.099673  0.154999  0.099964   
Age                       0.137139  0.026890  0.127312   

                          DiabetesPedigreeFunction       Age  
Pregnancies                              -0.004637  0.512118  
Glucose                                   0.137858  0.267034  
BloodPressure                            -0.001156  0.325326  
SkinThickness                             0.099964  0.127312  
Insulin                                   0.099673  0.137139  
BMI                                       0.154999  0.026890  
Skin                                      0.099964  0.127312  
DiabetesPedigreeFunction                  1.000000  0.034402  
Age                                       0.034402  1.000000

In [467]:
del df_pandas['Skin']

In [468]:
df_pandas.head(10)

Pregnancies  Glucose  BloodPressure  SkinThickness     Insulin        BMI  \
0     6.000000    148.0      72.000000      35.000000  155.379747  33.600000   
1     1.000000     85.0      66.000000      29.000000  155.379747  26.600000   
2     8.000000    183.0      64.000000      29.164207  155.379747  23.300000   
3     1.000000     89.0      66.000000      23.000000   94.000000  28.100000   
4     4.490881    137.0      40.000000      35.000000  168.000000  43.100000   
5     5.000000    116.0      74.000000      29.164207  155.379747  25.600000   
6     3.000000     78.0      50.000000      32.000000   88.000000  31.000000   
7    10.000000    115.0      72.384196      29.164207  155.379747  35.300000   
8     2.000000    197.0      70.000000      45.000000  543.000000  30.500000   
9     8.000000    125.0      96.000000      29.164207  155.379747  32.443404   

   DiabetesPedigreeFunction  Age DiabetesDiagnosis  
0                     0.627   50               yes  
1                     0.351   31                no  
2                     0.672   32               yes  
3                     0.167   21                no  
4                     2.288   33               yes  
5                     0.201   30                no  
6                     0.248   26               yes  
7                     0.134   29                no  
8                     0.158   53               yes  
9                     0.232   54               yes

### This Concludes the Data ETL, and ready to train the model  

In [469]:
df_spark = sqlContext.createDataFrame(df_pandas)

In [470]:
df_spark.show(10)

+-----------------+-------+----------------+-----------------+------------------+-----------------+------------------------+---+-----------------+
|      Pregnancies|Glucose|   BloodPressure|    SkinThickness|           Insulin|              BMI|DiabetesPedigreeFunction|Age|DiabetesDiagnosis|
+-----------------+-------+----------------+-----------------+------------------+-----------------+------------------------+---+-----------------+
|              6.0|  148.0|            72.0|             35.0|155.37974683544303|             33.6|                   0.627| 50|              yes|
|              1.0|   85.0|            66.0|             29.0|155.37974683544303|             26.6|     0.35100000000000003| 31|               no|
|              8.0|  183.0|            64.0|29.16420664206642|155.37974683544303|             23.3|                   0.672| 32|              yes|
|              1.0|   89.0|            66.0|             23.0|              94.0|             28.1|     0.166999999999

## 3. Create an Apache® Spark machine learning model
In this section you will learn how to prepare data, create and train an Apache® Spark machine learning model.

### 3.1: Prepare Training & Validation (Testing) data
In this subsection you will split your data into: train and  test  data sets.

In [471]:
split_data = df_spark.randomSplit([0.7, 0.3], 24)
train_data = split_data[0]
test_data = split_data[1]


print("Number of training records: " + str(train_data.count()))
print("Number of testing records : " + str(test_data.count()))

Number of training records: 538
Number of testing records : 231


As you can see our data has been successfully split into two data sets:
* The train data set, which is the largest group, is used for training.
* The test data set will be used for model evaluation and is used to test the assumptions of the model.

### 3.2: Create pipeline and train a model
In this section you will create an Apache® Spark machine learning pipeline and then train the model.
In the first step you need to import the Apache® Spark machine learning packages that will be needed in the subsequent steps.

A sequence of data processing is called a _data pipeline_.  Each step in the pipeline processes the data and passes the result to the next step in the pipeline, this allows you to transform and fit your model with the raw input data.

In [472]:
from pyspark.ml.feature import StringIndexer, IndexToString, VectorAssembler
from pyspark.ml.classification import RandomForestClassifier
from pyspark.ml.evaluation import MulticlassClassificationEvaluator
from pyspark.ml import Pipeline, Model

In the following step, convert all the string fields to numeric ones by using the StringIndexer transformer.

In [473]:
stringIndexer_label = StringIndexer(inputCol="DiabetesDiagnosis", outputCol="label").fit(df_data)


In the following step, create a feature vector by combining all features together.

In [474]:
vectorAssembler_features = VectorAssembler(inputCols=['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI', 'DiabetesPedigreeFunction', 'Age'], outputCol='features')

Next, define estimators you want to use for classification. Random Forest is used in the following example.

In [475]:
rf = RandomForestClassifier(labelCol="label", featuresCol="features")

Finally, indexed labels back to original labels.

In [476]:
labelConverter = IndexToString(inputCol="prediction", outputCol="predictedLabel", labels=stringIndexer_label.labels)

In [477]:
transform_df_pipeline = Pipeline(stages=[stringIndexer_label, vectorAssembler_features])
transformed_df = transform_df_pipeline.fit(df_spark).transform(df_spark)
transformed_df.show()

+-----------------+-------+----------------+-----------------+------------------+-----------------+------------------------+---+-----------------+-----+--------------------+
|      Pregnancies|Glucose|   BloodPressure|    SkinThickness|           Insulin|              BMI|DiabetesPedigreeFunction|Age|DiabetesDiagnosis|label|            features|
+-----------------+-------+----------------+-----------------+------------------+-----------------+------------------------+---+-----------------+-----+--------------------+
|              6.0|  148.0|            72.0|             35.0|155.37974683544303|             33.6|                   0.627| 50|              yes|  1.0|[6.0,148.0,72.0,3...|
|              1.0|   85.0|            66.0|             29.0|155.37974683544303|             26.6|     0.35100000000000003| 31|               no|  0.0|[1.0,85.0,66.0,29...|
|              8.0|  183.0|            64.0|29.16420664206642|155.37974683544303|             23.3|                   0.672| 32|  

Let's build the pipeline now. A pipeline consists of transformers and an estimator.

In [478]:
pipeline_rf = Pipeline(stages=[stringIndexer_label, vectorAssembler_features, rf, labelConverter])

Now, you can train your Random Forest model by using the previously defined **pipeline** and **training data**.

In [479]:
model_rf = pipeline_rf.fit(train_data)

You can check your **model accuracy** now. To evaluate the model, use **test data**.

In [480]:
predictions_Train = model_rf.transform(train_data)
evaluatorRF = MulticlassClassificationEvaluator(labelCol="label", predictionCol="prediction", metricName="accuracy")
accuracy_Train = evaluatorRF.evaluate(predictions_Train)
print("Accuracy_Train = %g" % accuracy_Train)
print("Test Error Train = %g" % (1.0 - accuracy_Train))

Accuracy_Train = 0.856877
Test Error Train = 0.143123


In [481]:
predictions = model_rf.transform(test_data)
evaluatorRF = MulticlassClassificationEvaluator(labelCol="label", predictionCol="prediction", metricName="accuracy")
accuracy = evaluatorRF.evaluate(predictions)
print("Accuracy Test = %g" % accuracy)
print("Test Error Test = %g" % (1.0 - accuracy))

Accuracy Test = 0.744589
Test Error Test = 0.255411


## 86% Accuracy meets the Business requirement (70% +).  

I also tried other couple of models such as Naive Bayes and LDA Models, which have slight lower Accuracy. Need to adjust iteratively with project progress

## 4. Persist model
Store pipeline and model in Watson Machine Learning repository by using Python client libraries.

In [482]:
#import client libraries.
from repository.mlrepositoryclient import MLRepositoryClient
from repository.mlrepositoryartifact import MLRepositoryArtifact

Authenticate to Watson Machine Learning service on IBM Cloud.

### 4.1 **Prepare to Deploy** 
Authentication information (username, password, and instance_id)  from the instance of Watson Machine Learning service

In [483]:
#Specify your username, password, and instance_id credientials for Watson ML
service_path = 'https://us-south.ml.cloud.ibm.com'
username = '1a1a287f-3d4c-4445-aa21-d683799704a7'
password = '5a2f9aba-9fb9-490b-99db-6434f889b31c'
instance_id = '85ed9882-9243-4cf1-8a30-30702d99fab4'

**Tip:** service_path, username, password, and instance_id can be found on Service Credentials tab of the Watson Machine Learning service instance created on the IBM Cloud.

In [484]:
ml_repository_client = MLRepositoryClient(service_path)
ml_repository_client.authorize(username, password)

Create model artifact (abstraction layer).

In [485]:
pipeline_artifact = MLRepositoryArtifact(pipeline_rf, name="pipeline")

model_artifact = MLRepositoryArtifact(model_rf, training_data=train_data, name="Diabetes Prediction Model", pipeline_artifact=pipeline_artifact)

### 4.2: Save pipeline and model¶
Save pipeline and model artifacts to the Watson Machine Learning instance.

In [486]:
saved_model = ml_repository_client.models.save(model_artifact)

Get saved model metadata from Watson Machine Learning.
**Tip:** Use *meta.availableProps* to get the list of available props.

In [487]:
saved_model.meta.available_props()

dict_keys(['trainingDataSchema', 'lastUpdated', 'inputDataSchema', 'evaluationMetrics', 'version', 'pipelineType', 'modelVersionHref', 'evaluationMethod', 'trainingDataRef', 'label', 'pipelineVersionHref', 'authorEmail', 'creationTime', 'modelType', 'authorName', 'runtime'])

In [488]:
print("modelType: " + saved_model.meta.prop("modelType"))
print("trainingDataSchema: " + str(saved_model.meta.prop("trainingDataSchema")))
print("creationTime: " + str(saved_model.meta.prop("creationTime")))
print("modelVersionHref: " + saved_model.meta.prop("modelVersionHref"))
print("label: " + saved_model.meta.prop("label"))

modelType: sparkml-model-2.1
trainingDataSchema: {'type': 'struct', 'fields': [{'metadata': {}, 'nullable': True, 'type': 'double', 'name': 'Pregnancies'}, {'metadata': {}, 'nullable': True, 'type': 'double', 'name': 'Glucose'}, {'metadata': {}, 'nullable': True, 'type': 'double', 'name': 'BloodPressure'}, {'metadata': {}, 'nullable': True, 'type': 'double', 'name': 'SkinThickness'}, {'metadata': {}, 'nullable': True, 'type': 'double', 'name': 'Insulin'}, {'metadata': {}, 'nullable': True, 'type': 'double', 'name': 'BMI'}, {'metadata': {}, 'nullable': True, 'type': 'double', 'name': 'DiabetesPedigreeFunction'}, {'metadata': {}, 'nullable': True, 'type': 'long', 'name': 'Age'}, {'metadata': {}, 'nullable': True, 'type': 'string', 'name': 'DiabetesDiagnosis'}]}
creationTime: 2019-01-13 21:03:34.902000+00:00
modelVersionHref: https://us-south.ml.cloud.ibm.com/v2/artifacts/models/943ad7ea-80dc-4c25-8f69-acde298da850/versions/63caec3d-ffeb-4d7c-86ac-dda6aa21f5c9
label: DiabetesDiagnosis



## 4.3 Load model to verify that it was saved correctly
You can load your model  to make sure that it was saved  correctly.

In [489]:
loadedModelArtifact = ml_repository_client.models.get(saved_model.uid)

Print the  model name to make sure that model artifact has been loaded correctly.

In [490]:
print(str(loadedModelArtifact.name))

Diabetes Prediction Model


## Should be sucessfully created the model and saved it in the Watson Machine Learning service. 
Now continue within the notebook to deploy the model using the APIs. (Alternatively, can use the Watson ML Console

***
***

## 6.0 Accessing Watson ML Models and Deployments through API
Instead of jumping from your notebook into a web browser, manage your model and delopment through a set of APIs


Recap of saving an existing ML model through using the Watson-Machine-Learning Python SDK


`pip install watson-machine-learning-client`

[SDK Documentation](https://watson-ml-libs.mybluemix.net/repository-python/index.html)

### Save model to WML Service

In [491]:
#Import Python WatsonML Repository SDK
from repository.mlrepositoryclient import MLRepositoryClient
from repository.mlrepositoryartifact import MLRepositoryArtifact

#Authenticate
ml_repository_client = MLRepositoryClient(service_path)
ml_repository_client.authorize(username, password)

#Deploy a new model.  I renamed the existing model as it has already been created above
pipeline_artifact = MLRepositoryArtifact(pipeline_rf, name="pipeline")

model_artifact = MLRepositoryArtifact(model_rf, training_data=train_data, name="Diabetes Prediction Model", pipeline_artifact=pipeline_artifact)

### 6.1 Get the Watson ML API Token
The Watson ML API authenticates all requests through a token, start by requesting the token from our Watson ML Service.

In [492]:
import json
import requests
from base64 import b64encode

token_url = service_path + "/v3/identity/token"

# NOTE: for python 2.x, uncomment below, and comment out the next line of code:
#userAndPass = b64encode(bytes(username + ':' + password)).decode("ascii")
# Use below for python 3.x, comment below out for python 2.x
userAndPass = b64encode(bytes(username + ':' + password, "utf-8")).decode("ascii")
headers = { 'Authorization' : 'Basic %s' %  userAndPass }

response = requests.request("GET", token_url, headers=headers)

watson_ml_token = json.loads(response.text)['token']
print(watson_ml_token)

eyJhbGciOiJSUzUxMiIsInR5cCI6IkpXVCJ9.eyJ0ZW5hbnRJZCI6Ijg1ZWQ5ODgyLTkyNDMtNGNmMS04YTMwLTMwNzAyZDk5ZmFiNCIsImluc3RhbmNlSWQiOiI4NWVkOTg4Mi05MjQzLTRjZjEtOGEzMC0zMDcwMmQ5OWZhYjQiLCJwbGFuSWQiOiIzZjZhY2Y0My1lZGU4LTQxM2EtYWM2OS1mOGFmM2JiMGNiZmUiLCJyZWdpb24iOiJ1cy1zb3V0aCIsInVzZXJJZCI6IjFhMWEyODdmLTNkNGMtNDQ0NS1hYTIxLWQ2ODM3OTk3MDRhNyIsImlzcyI6Imh0dHBzOi8vdXMtc291dGgubWwuY2xvdWQuaWJtLmNvbS92My9pZGVudGl0eSIsImlhdCI6MTU0NzQxMzQ0OSwiZXhwIjoxNTQ3NDQyMjQ5LCJjcmVhdGVkVGltZSI6MTU0NzQxMzQ0OX0.QFksVPQ6J7q8rJkaM5hZi2E43Ry9f0dY9RYdOgwRFJx2xBFBl7xI0w5g-b9zOTTv8zOmQ9_DEOpm5fs3KajUjRIirkImpK3GQ2CchZfrpTN6eyGxNdhC6cd0S78_6xrIb8BigvE7KJwFO5ZA6PIXQpaE5Jiw7sS6x1aFPHONWhiYeG7cyYczkrnBJtppMgW5C7eQM9DjPIgQfFvzfub_7Q2zof-gCDucI4fkSoKpKl_ZbvD4dZRKt8dcsrnyI3L9-J0ChXo1pEa1N-LZAPxdQtcZTcr5CRMUXHLTGzpWl9FlnS-RZtQFKHjwoNQ-dyGidqnEnOAJrxaQH3f5RHK44A


### 6.2 Preview currenly published models

In [493]:
model_url = service_path + "/v3/wml_instances/" + instance_id + "/published_models"

headers = {'authorization': 'Bearer ' + watson_ml_token }
response = requests.request("GET", model_url, headers=headers)

published_models = json.loads(response.text)
print(json.dumps(published_models, indent=2))

{
  "limit": 1000,
  "first": {
    "url": "https://us-south.ml.cloud.ibm.com/v3/wml_instances/85ed9882-9243-4cf1-8a30-30702d99fab4/published_models?limit=1000"
  },
  "resources": [
    {
      "metadata": {
        "created_at": "2019-01-13T02:22:46.797Z",
        "url": "https://us-south.ml.cloud.ibm.com/v3/wml_instances/85ed9882-9243-4cf1-8a30-30702d99fab4/published_models/b48c5a26-7e03-44d8-9e32-5bfd0b32fe7e",
        "modified_at": "2019-01-13T20:52:19.329Z",
        "guid": "b48c5a26-7e03-44d8-9e32-5bfd0b32fe7e"
      },
      "entity": {
        "latest_version": {
          "created_at": "2019-01-13T02:22:46.873Z",
          "url": "https://us-south.ml.cloud.ibm.com/v3/ml_assets/models/b48c5a26-7e03-44d8-9e32-5bfd0b32fe7e/versions/a8f88952-5102-48fa-b559-146514b7b876",
          "guid": "a8f88952-5102-48fa-b559-146514b7b876"
        },
        "runtime_environment": "spark-2.1",
        "name": "Heart Failure Prediction Model",
        "learning_configuration_url": "https://us

Read the details of any returned models

In [495]:
print('{} model(s) are available in your Watson ML Service'.format(len(published_models['resources'])))
for model in published_models['resources']:
    print('\t- name:        {}'.format(model['entity']['name']))
    print('\t  model_id:    {}'.format(model['metadata']['guid']))
    print('\t  deployments: {}'.format(model['entity']['deployments']['count']))

12 model(s) are available in your Watson ML Service
	- name:        Heart Failure Prediction Model
	  model_id:    b48c5a26-7e03-44d8-9e32-5bfd0b32fe7e
	  deployments: 0
	- name:        Heart Failure Prediction Model
	  model_id:    665d51ae-de63-4e4a-8216-c8188d035126
	  deployments: 0
	- name:        Diabetes Prediction Model
	  model_id:    91e0299a-b03d-45ba-9e10-a866df2abbe8
	  deployments: 0
	- name:        Diabetes Prediction Model
	  model_id:    9b5d1e16-acbc-4d50-9085-4593e08ad4af
	  deployments: 0
	- name:        Diabetes Prediction Model
	  model_id:    41dd2272-05b8-4728-bd20-e2f7f56392bb
	  deployments: 0
	- name:        Diabetes Prediction Model
	  model_id:    8059d3db-e818-4a46-b841-c3647c2c17bc
	  deployments: 0
	- name:        Diabetes Prediction Model
	  model_id:    416df992-f3cf-4d81-a30f-df2d7597c439
	  deployments: 0
	- name:        Diabetes Prediction Model
	  model_id:    70a3ffed-bf59-4916-ba72-51ecf95ed877
	  deployments: 0
	- name:        Diabetes Predictio

Create a new deployment of the Model

In [506]:

# Update this `model_id` with the model_id from model that you wish to deploy listed above.
model_id = '91e0299a-b03d-45ba-9e10-a866df2abbe8'

deployment_url = service_path + "/v3/wml_instances/" + instance_id + "/published_models/" + model_id + "/deployments"

payload = "{\"name\": \"Diabetes Prediction Model Deployment\", \"description\": \"First deployment of Diabetes Prediction Model\", \"type\": \"online\"}"
headers = {'authorization': 'Bearer ' + watson_ml_token, 'content-type': "application/json" }

response = requests.request("POST", deployment_url, data=payload, headers=headers)

print(response.text)

{
  "metadata": {
    "guid": "b405cd57-24e9-4afb-99e8-240f8bc0e532",
    "url": "https://us-south.ml.cloud.ibm.com/v3/wml_instances/85ed9882-9243-4cf1-8a30-30702d99fab4/published_models/91e0299a-b03d-45ba-9e10-a866df2abbe8/deployments/b405cd57-24e9-4afb-99e8-240f8bc0e532",
    "created_at": "2019-01-13T21:08:58.402Z",
    "modified_at": "2019-01-13T21:09:07.753Z"
  },
  "entity": {
    "runtime_environment": "spark-2.1",
    "name": "Diabetes Prediction Model Deployment",
    "scoring_url": "https://us-south.ml.cloud.ibm.com/v3/wml_instances/85ed9882-9243-4cf1-8a30-30702d99fab4/published_models/91e0299a-b03d-45ba-9e10-a866df2abbe8/deployments/b405cd57-24e9-4afb-99e8-240f8bc0e532/online",
    "deployable_asset": {
      "name": "Diabetes Prediction Model",
      "url": "https://us-south.ml.cloud.ibm.com/v3/wml_instances/85ed9882-9243-4cf1-8a30-30702d99fab4/published_models/91e0299a-b03d-45ba-9e10-a866df2abbe8",
      "guid": "91e0299a-b03d-45ba-9e10-a866df2abbe8",
      "created_at": "

In [507]:
deployment = json.loads(response.text)

print('Model {} deployed.'.format(model_id))
print('\tname: {}'.format(deployment['entity']['name']))
print('\tdeployment_id: {}'.format(deployment['metadata']['guid']))
print('\tstatus: {}'.format(deployment['entity']['status']))
print('\tscoring_url: {}'.format(deployment['entity']['scoring_url']))

Model 91e0299a-b03d-45ba-9e10-a866df2abbe8 deployed.
	name: Diabetes Prediction Model Deployment
	deployment_id: b405cd57-24e9-4afb-99e8-240f8bc0e532
	status: DEPLOY_SUCCESS
	scoring_url: https://us-south.ml.cloud.ibm.com/v3/wml_instances/85ed9882-9243-4cf1-8a30-30702d99fab4/published_models/91e0299a-b03d-45ba-9e10-a866df2abbe8/deployments/b405cd57-24e9-4afb-99e8-240f8bc0e532/online


### Monitor the status of deployment

In [508]:

# Update this `deployment_id` from the newly deployed model from above.
deployment_id = "3998269c-f9ff-4a9b-b391-83cc59e5c09a"
deployment_details_url = service_path + "/v3/wml_instances/" + instance_id + "/published_models/" + model_id + "/deployments/" + deployment_id

headers = {'authorization': 'Bearer ' + watson_ml_token, 'content-type': "application/json" }

response = requests.request("GET", deployment_url, headers=headers)
print(response.text)

{
  "count": 2,
  "resources": [{
    "metadata": {
      "guid": "3998269c-f9ff-4a9b-b391-83cc59e5c09a",
      "url": "https://us-south.ml.cloud.ibm.com/v3/wml_instances/85ed9882-9243-4cf1-8a30-30702d99fab4/deployments/3998269c-f9ff-4a9b-b391-83cc59e5c09a",
      "created_at": "2019-01-13T21:04:35.742Z",
      "modified_at": "2019-01-13T21:04:38.982Z"
    },
    "entity": {
      "runtime_environment": "spark-2.1",
      "name": "Diabetes Prediction Model Deployment",
      "scoring_url": "https://us-south.ml.cloud.ibm.com/v3/wml_instances/85ed9882-9243-4cf1-8a30-30702d99fab4/deployments/3998269c-f9ff-4a9b-b391-83cc59e5c09a/online",
      "deployable_asset": {
        "name": "Diabetes Prediction Model",
        "url": "https://us-south.ml.cloud.ibm.com/v3/wml_instances/85ed9882-9243-4cf1-8a30-30702d99fab4/published_models/91e0299a-b03d-45ba-9e10-a866df2abbe8",
        "guid": "91e0299a-b03d-45ba-9e10-a866df2abbe8",
        "created_at": "2019-01-13T06:08:41.567Z",
        "type": "mo

In [509]:
deployment_details = json.loads(response.text)

for resources in deployment_details['resources']:
    print('name: {}'.format(resources['entity']['name']))
    print('status: {}'.format(resources['entity']['status']))
    print('scoring url: {}'.format(resources['entity']['scoring_url']))

name: Diabetes Prediction Model Deployment
status: DEPLOY_SUCCESS
scoring url: https://us-south.ml.cloud.ibm.com/v3/wml_instances/85ed9882-9243-4cf1-8a30-30702d99fab4/deployments/3998269c-f9ff-4a9b-b391-83cc59e5c09a/online
name: Diabetes Prediction Model Deployment
status: DEPLOY_SUCCESS
scoring url: https://us-south.ml.cloud.ibm.com/v3/wml_instances/85ed9882-9243-4cf1-8a30-30702d99fab4/deployments/b405cd57-24e9-4afb-99e8-240f8bc0e532/online


## 6.3 Invoke prediction model deployment
Define a method to call scoring url. Replace the **scoring_url** in the method below with the scoring_url returned from above.

In [510]:
def get_prediction_ml(pregnancies,glucose,bloodpressure,skinthickness,insulin,bmi,diabetesPedigreeFunction,age):
    scoring_url = 'https://us-south.ml.cloud.ibm.com/v3/wml_instances/85ed9882-9243-4cf1-8a30-30702d99fab4/deployments/3998269c-f9ff-4a9b-b391-83cc59e5c09a/online'
    scoring_payload = { "fields":['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI', 'DiabetesPedigreeFunction', 'Age'],"values":[[pregnancies,glucose,bloodpressure,skinthickness,insulin,bmi,diabetesPedigreeFunction,age]]}
    header = {'authorization': 'Bearer ' + watson_ml_token, 'content-type': "application/json" }
    scoring_response = requests.post(scoring_url, json=scoring_payload, headers=header)
    return (json.loads(scoring_response.text).get("values")[0][12])

### Call get_prediction_ml method exercising our prediction model

In [511]:
print('Is a 50 year old female that smokes with a low BMI at risk of Diabetes?: {}'.format(get_prediction_ml(6,148,72,35,0,33.6,0.627,50)))

Is a 50 year old female that smokes with a low BMI at risk of Diabetes?: yes


In [512]:
print('Is a 31 year old female that smokes with a low BMI at risk of Diabetes?: {}'.format(get_prediction_ml(1,85,66,29,0,26.6,0.351,31)))

Is a 31 year old female that smokes with a low BMI at risk of Diabetes?: no
